src/governance-validators.cpp

```diff
@@ -0,0 +1,349 @@
+        strErrorMessages += "name contains invalid characters;";
+        return false;
+    }
+
+    return true;
+}
+
+bool CProposalValidator::ValidateStartEndEpoch()
+{
+    int64_t nStartEpoch = 0;
+    int64_t nEndEpoch = 0;
+
+    if(!GetDataValue("start_epoch", nStartEpoch)) {
+        strErrorMessages += "start_epoch field not found;";
+        return false;
+    }
+
+    if(!GetDataValue("end_epoch", nEndEpoch)) {
+        strErrorMessages += "end_epoch field not found;";
+        return false;
+    }
+
+    if(nEndEpoch <= nStartEpoch) {
+        strErrorMessages += "end_epoch <= start_epoch field not found;";
+        return false;
+    }
+
+    return true;
+}
+
+bool CProposalValidator::ValidatePaymentAmount()
+{
+    double dValue = 0.0;
+
+    if(!GetDataValue("payment_amount", dValue)) {
+        strErrorMessages += "payment_amount field not found;";
+        return false;
+    }
+
+    if(dValue <= 0.0) {
+        strErrorMessages += "payment_amount invalid;";
+        return false;
+    }
+
+    return true;
+}
+
+bool CProposalValidator::ValidatePaymentAddress()
+{
+    std::string strPaymentAddress;
+
+    if(!GetDataValue("payment_address", strPaymentAddress)) {
+        strErrorMessages += "payment_address field not found;";
+        return false;
+    }
+
+    static const std::string base58chars = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";
+
+    size_t nLength = strPaymentAddress.size();
+
+    if((nLength < 26) || (nLength > 35)) {
+        strErrorMessages += "incorrect payment_address length;";
+        return false;
+    }
+
+    if(strPaymentAddress.find_first_not_of(base58chars) != std::string::npos) {
+        strErrorMessages += "payment_address contains invalid characters;";
+        return false;
+    }
+
+    CBitcoinAddress address(strPaymentAddress);
+    if(!address.IsValid()) {
+        strErrorMessages += "payment_address is invalid;";
+        return false;
+    }
+
+    return true;
+}
+
+bool CProposalValidator::ValidateURL()
+{
+    std::string strURL;
+    if(!GetDataValue("url", strURL)) {
+        strErrorMessages += "url field not found;";
+        return false;
+    }
+
+    std::string strURLStripped = StripWhitespace(strURL);
+
+    if(int(strURLStripped.size()) < 4) {
+        strErrorMessages += "url too short;";
+        return false;
+    }
+
+    if(!CheckURL(strURL)) {
+        strErrorMessages += "url invalid;";
+        return false;
+    }
+
+    return true;
+}
+
+void CProposalValidator::ParseJSONData()
+{
+    fJSONValid = false;
+
+    if(strData.empty()) {
+        return;
+    }
+
+    try {
+        UniValue obj(UniValue::VOBJ);
+        obj.read(strData);
+        std::vector<UniValue> arr1 = obj.getValues();
+        std::vector<UniValue> arr2 = arr1.at(0).getValues();
+        objJSON = arr2.at(1);
+        fJSONValid = true;
+    }
+    catch(std::exception& e) {
+        strErrorMessages += std::string(e.what()) + std::string(";");
```

Comment:
More explanation on my comment.
If the error message type is changed from string to stringstream then it is more clearly redundant to convert output `char*` parameters to strings before streaming them out:
```
strstreamErrorMessages << e.what() << ";";
```
Apart from type and syntax, the calling pattern is the same so there is no reason to treat the parameters differently in string vs stringstream (except that the stream is more flexible - the semicolon can be output as a character literal instead of a string literal)
Sticking with string, the two string conversions on the output can be dropped:
```
Consider using stringstream as a more suitable type for concatenating outputs.
strErrorMessages = strErrorMessages + e.what() + ";";
```